pilot/pkg/proxy/envoy/v2/mesh.go

```diff
@@ -0,0 +1,990 @@
+				}
+				newEpSubset := m.reverseAttrMap.getEndpointsMatching(labelsToMatch)
+				if ruleType == DestinationRuleCIDR || ruleType == DestinationRuleWildcard {
+					// For rule names that do not map to values that can be label matched
+					// scope the returned set to the type of rule.
+					newEpSubset = newEpSubset.scopeToRule(ruleType, labelValue, cidrNet)
+				}
+				m.subsetEndpoints.addEndpointSet(scopedSSName, newEpSubset)
+				m.reverseEpSubsets.addSubsetForEndpoints(newEpSubset, scopedSSName)
+				m.subsetDefinitions[scopedSSName] = subset
+			}
+		}
+	}
+	return errs
+}
+
+// NewEndpoint is a boiler plate function intended for platform Controllers to create a new Endpoint.
+// This method ensures all the necessary data required for creating subsets are correctly setup. It
+// also performs sorting of arrays etc, to allow stable results for reflect.DeepEquals() for quick
+// comparisons. The address	is the network address of this endpoint that must be accessible from __THIS__
+// Pilot, i.e. one importing the Endpoint into its Mesh via Reconcile(). If a remote exposes a gateway,
+// the remote pilot exposes the gateway's address. The gateway itself may have more than one Endpoints
+// behind it that are not directly network accessible from this Pilot. Similarly the the network port
+// of this endpoint that must be accessible from __THIS__ pilot. socketProtocol should be set to
+// TCP or UPD. Labels are properties of the workload, for example: pod labels in Kubernetes.
+func NewEndpoint(address string, port uint32, socketProtocol SocketProtocol, labels []EndpointLabel) (*Endpoint, error) {
+	var errs error
+	ipAddr := net.ParseIP(address)
+	if ipAddr == nil {
+		errs = multierror.Append(errs, fmt.Errorf("invalid IP address '%s'", address))
+	}
+	hasUID := false
+	svcFQDNs := []string{}
+	svcDomains := []string{}
+	miscLabels := make(map[string]string, len(labels))
+	for _, label := range labels {
+		if prohibitedAttrSet[label.Name] {
+			errs = multierror.Append(fmt.Errorf(
+				"prohibited use of Istio destination label '%s' for endpoint label", label.Name))
+			continue
+		}
+		switch label.Name {
+		case DestinationService.AttrName():
+			svcFQDNs = append(svcFQDNs, label.Value)
+		case DestinationDomain.AttrName():
+			svcDomains = append(svcDomains, label.Value)
+		case DestinationUID.AttrName():
+			hasUID = true
+			fallthrough
+		default:
+			oldValue, found := miscLabels[label.Name]
+			if found {
+				errs = multierror.Append(fmt.Errorf(
+					"single value label '%s' has multiple values ['%s','%s']", label.Name, oldValue, label.Value))
+				continue
+			}
+			miscLabels[label.Name] = label.Value
+		}
+	}
+	if !hasUID {
+		errs = multierror.Append(fmt.Errorf(
+			"missing Endpoint mandatory label '%s'", DestinationUID.AttrName()))
+	}
+	if len(svcFQDNs) == 0 {
+		errs = multierror.Append(fmt.Errorf(
+			"missing Endpoint mandatory label '%s'", DestinationService.AttrName()))
+	}
+	if errs != nil {
```

Comment:
You've done this pattern in a few different places. As written, every time this appears its safe, but just a word of warning: even if you have `nil` errors, a round trip through `multierror` makes them non-nil (an empty multierror is still non-nil). Now, because any occasion that you call `multierror.Append` you've for sure got an error, `errs` will only be non-nil if there was an error, but this could trip people up later on.

We should move this helper method up to `pkg/errors` or similar, I think. https://github.com/istio/istio/blob/6cae0b991ad73742518914d95cdf7ac6f1060b1a/pilot/pkg/model/validation.go#L2099-L2115

Reply:
done